[ADD] pos_online_payment, *: add online payments

thvi-odoo opened the pull request:
*: point_of_sale, pos_adyen, pos_hr, pos_mercury, pos_restaurant

Adds online payments in Point of Sale using the payment module.
The pos_online_payment module is meant to be used by shops for paying online instead of using a physical terminal and by restaurants for paying an order either from the paying station or in the self-order process (smartphone used to order and pay online).
The module depends only on point_of_sale (and account_payment to support online payments).
The online payments support for self-order will be added in another commit.
The feature can be used by configuring a special pos.payment.method for one or many pos.config.

Operation:
The frontend session user can enter one or several online payment method lines (with or without other payment methods).
When he validates the order, if the online payment amounts are valid, a popup with a QR code is displayed on the frontend screen and on the customer display. This QR code allows the customer to open the payment page to pay online for the order.

The amount to pay for an order on the payment portal is the remaining unpaid amount or a lower amount if several online payments lines are used. Therefore, at any time, the customer will be able to pay the amount of the order that is unpaid, without requiring a staff operation to update the amount to pay.

When an online payment is done for a POS order, the generated account.payment is linked to the POS order through a pos.payment that is created with the online payment method configured for the pos.config, and the POS order paid process is triggered (it is the responsibility of the server to finalize the validation of POS orders that have at least one online payment, and not the responsibility of the POS session web client). If for any reason, there is no longer any online payment method configured for the pos.config of the order, a new one can be created automatically and exceptionnaly used to link the done payment with the POS order. This guarantees that any online payment saved by the payment module will be linked to the POS order.

The session web client is notified of successful online payments through web socket communication, to make the cashier screen automatically and quickly react to payments.

If several online payments are used for a single POS order, they are processed one after each other.

Working principles:
 - The POS session web client is not allowed to save online payments on the server. Only the Odoo server, which handles the payment, can save online payments for a POS order.
 - The POS session is only responsible for saving the order (especially other payment methods) and the next online payment amount to pay (in case of several online payments) on the server before the online payments are made by the customer, and for displaying the online payment status and the link (QR code) to open the payment page.
That means that modifications that are made after the order is fully paid are not taken into account by the server.
 - Before validating an order, the POS session checks with the server if any online payment has been made for that POS order. If the user wasn't aware of one, the validation process is aborted, notifying him.
 - The sensitive online payments data is received by the frontend user with RPC to the server when validating an order, cancelling an online payment or receiving a notification through the POS bus (web socket communication).
This data is used to update the cashier data for the specific order.
 - No sensitive information is sent through the POS bus (web socket communication).
Indeed, the bus communication is only protected by the name of the channel, and there is no mecanism to prevent several attempts. Therefore, no sensitive information is sent through it, only a notification to invite the local browser to do a safe RPC to the server to check the new state of the order.
 - Online payments for a POS order cannot be modified or removed when they succeeded. Checks ensure to never allow an online payment to make the total paid amount greater than the total amount of the POS order.
 - If the frontend user clicks on the "Cancel" button of the online payment popup, the order can potentially still be paid online. Indeed, the order will be removed from the server only if it was added only for the online payment flow, and any order that is not paid can be paid on the online payment portal (if the customer has the web link).
This behavior ensures that in a shop, starting an online payment and then cancelling it doesn't prevent the user to close the POS session. And it ensures to don't delete a draft order that is used for other flows, like in a restaurant or when there are trusted POS configs.

Payment portal features:
 - The payment portal pages of a POS order are protected with its access_token.
 - If the amount of the order has changed when the customer clicks to pay, an error is fired to notify him and give him the choice to continue or abandon the payment.
 - An exit_route argument (an URL) can be passed to allow the customer to exit the payment portal at any time (with a button), which could be used for self-order flow.
 - A button to go back to the payment page is displayed when an error occurs during a payment transaction, which allows the customer to retry a payment without needing to scan again the QR code.
 - If the customer is not logged in in the payment portal (most frequent case), it is considered as the public user partner for the payment flow to don't give him privileged access to sensitive data like payment tokens.
 - The accounting process of online payments for POS orders is made in the same way that the pos.payment.method that have split_transactions to True.
 - The POS customer of a POS order has no link with the customer making an online payment for that POS order.
 - The payment portal is made using the pos.order id and not the pos_reference because the latter is not guaranteed to be unique.

Some tests have been written, especially one that simulates a whole online payment flow starting from nothing, creating an order, requesting an online payment, faking this online payment by doing web requests to the payment portal, then finally closing the POS session to check that there is no accounting issue.

The design of the payment portal is temporary.

task-id: 3276404
related: https://github.com/odoo/enterprise/pull/41776

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/123237).

caburj commented on addons/point_of_sale/static/src/app/store/models.js:
```diff
@@ -1431,7 +1431,10 @@ export class Order extends PosModel {
         paymentLines = [];
         this.paymentlines.forEach((item) => {
-            return paymentLines.push([0, 0, item.export_as_JSON()]);
+            const itemAsJson = item.export_as_JSON();
+            if (itemAsJson) {
+                return paymentLines.push([0, 0, itemAsJson]);
+            }
```
Can you try again without this diff? You might have encountered this because of how the reactivity works. When removing items in a collection (array), the item is removed but the length is not updated, so there is a time where the an `undefined` slot is present in the array. This can be avoided if all the reactive effects we have is "batched".

thvi-odoo replied:
No this is because I return null for online payment lines here: https://github.com/odoo/odoo/pull/123237/files/0f62f4170488cdfc1439ef49452d971c3386ec93#diff-bdcc8aaeaf691939f64fe9da597375bb5145f498c84ecbc487766d489e11fde0R94-R111
This working principle can seem strange, but it is the easiest way, in my opinion, of never sending online payment lines to the server in the current state of the code. Changing this would require to change a lot of things in the way orders are sent to the server.

caburj replied:
Oh. Understood.

caburj commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -148,7 +148,12 @@ def _process_order(self, order, draft, existing_order):
         self._process_payment_lines(order, pos_order, pos_session, draft)
+        return pos_order._process_saved_order(draft)
 
+    @api.model
```
```suggestion
```

remove this. @api.model is analogous to @staticmethod. meaning, self should just be the model, not the records.

thvi-odoo replied:
Thank you, I didn't know about that.

caburj commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -165,6 +170,10 @@ def _process_order(self, order, draft, existing_order):
 
+    def _clean_payment_lines(self):
+        self.ensure_one()
+        order_bank_statement_lines = self.env['pos.payment'].search([('pos_order_id', '=', self.id)])
+        order_bank_statement_lines.unlink()
```
```suggestion
        order_payments = self.env['pos.payment'].search([('pos_order_id', '=', self.id)])
        order_payments.unlink()
```

let's take this opportunity to rename the variable.

caburj commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -574,6 +585,10 @@ def _create_invoice(self, move_vals):
                 })
         return new_move
 
+    def get_amount_unpaid(self):
+        self.ensure_one()
+        return self.currency_id.round(self._get_rounded_amount(self.amount_total) - self.amount_paid)
+
```
move this to pos_online_payment, unless you can find a place here in point_of_sale to call it.

caburj commented on addons/point_of_sale/static/src/app/customer_display/customer_display_template.xml:
Find a way to put this diff in the pos_online_payment addon. We should xpath an element here from there. The idea is to pretend that point_of_sale doesn't know about pos_online_payment addon.

thvi-odoo replied:
I did it changing a little bit the xml file of point_of_sale to make the things more clear.

caburj reviewed (commented):
Some actionable comments.

caburj commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+            raise ValidationError(_("The order has been canceled."))
+        return order_sudo
+
+    def _check_is_session_open(self, pos_order_sudo):
```
```suggestion
    def _ensure_session_open(self, pos_order_sudo):
```
rename

caburj commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+        amount_to_pay = self._get_amount_to_pay(pos_order_sudo)
+        if not self._is_valid_amount(amount_to_pay, currency_id):
+            rendering_context['amount'] = False
+            return self._render_pay(request, rendering_context)
```
What does it mean to call _render_pay with amount=False?

thvi-odoo replied:
There will be an error message, see addons/payment/views/payment_portal_templates.xml

caburj commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+
+    @staticmethod
+    def _render_pay(request, rendering_context):
+        return request.render('pos_online_payment.pay', rendering_context)
```
In this file, is there a way to generate custom `request` object? If all the `request` object here is the same, we can consider it global. No need to pass it around method/functions.

thvi-odoo replied:
I don't really know how that "request" object works, so I don't know if it is safe to use it in a staticmethod. For me, it makes sense that the "request" variable is linked to the current request of an "@http.route" method, but it doesn't make sense that the "request" variable would be linked to the correct request when the "@staticmethod" would be called. But it's only my assumption, I don't know well Python nor the Odoo API.

thvi-odoo replied:
Ok so I assume it is safe to call it from an @http.route method and use request in it without passing it as a parameter.
This method avoids copy/paste. If tomorrow the name of the "pay" template changes, or the way the context is passed to it is changed, only this method would need to be changed.

caburj commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+        if not payment_option_id:
+            raise UserError(_("A payment option must be specified."))
+        flow = kwargs.get('flow', None)
+        if not flow or flow not in ['redirect', 'direct', 'token']:
```
```suggestion
        if not (flow and flow in ['redirect', 'direct', 'token']):
```
I find it more readable this way. But if it's the opposite for you, no need to take this suggestion.

thvi-odoo replied:
I prefer my first approach but it's a matter of point of view and habits.

caburj commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+
+        tx_sudo = request.env['payment.transaction'].sudo().browse(tx_id)
```
```suggestion
        tx_sudo = request.env['payment.transaction'].sudo().search([('id', '=', tx_id)])
```
When you use `browse`, you probably need to call `.exists()` on it. E.g., you can do "browse(99999)" and it will just immediately create the object, without checking the db. (Something I observed recently.)

That's probably the reason why you need to handle `MissingError` below.

thvi-odoo replied:
Ok thank you, changed it

caburj reviewed (commented):
Some actionable comments on the payment_portal.py.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,302 @@
+            return self._render_pay(request, rendering_context)
+        rendering_context['amount'] = amount_to_pay
+
+        providers_sudo = self._get_allowed_providers_sudo(
```
Yes it is possible for different reasons (the user changes the payment method providers after the session has been opened, the payment providers configured are not compatible with this payment for any reason...).
In that case, we have the default "No suitable payment option could be found. If you believe that it is an error, please contact the website administrator." error message (from the payment template) (just tested).

caburj commented on addons/pos_online_payment/models/pos_session.py:
I guess my only comment here is that the "Customer" set to the order is different from the "partner_id" assigned to the move lines. Can you try to adjust the payment processing to use the partner set in the order?

thvi-odoo replied:
The problem is that the partner_id assigned to the move lines is the one used on the payment portal. If that partner is the one of the order, then it would mean that anyone who opens the payment portal page (scans the QR code, anyone can do it, there is not a box in shops/restaurants to only display the QR code to the right customer) would be authenticated as the customer of the POS order, and would have access to his payment tokens (saved payment credentials) for example.
This is a serious security issue in my opinion (we could have dishonest staff that could abuse of that security breach to fake their sales (the sales would be with real money, but without the customer consent) and earn a bonus from their boss for example).
Added to that, in the case of split payments, it doesn't make sense to have the same partner for payments done by 2 different persons for the same order.
In my opinion, "Public user" is the best approach when the customer is not authenticated (which is the most probable use) because it highlights clearly the reality: the user is not authenticated.

caburj replied:
Okay, we'll keep the use of public user if the customer doesn't authenticate.

caburj reviewed (commented):
Not much in the models. Except the partner_id of the pos.order doesn't match the partner_id in the payment.

thvi-odoo commented:
Hello @odoo/rd-security, theses changes are made to add online payments in the Point of Sale app using the 'payment' module of Odoo.
Here is an explanation for every triggered check security:
* The first 3 triggers:
    * [odoo/addons/pos_online_payment/controllers/payment_portal.py:4](https://github.com/odoo/odoo/blob/f033f94a029bc0455380259f1a7fc8489dc863e7/addons/pos_online_payment/controllers/payment_portal.py#L4)
import urllib.parse
Line may contain security risk: urllib.
    * [odoo/addons/pos_online_payment/controllers/payment_portal.py:39](https://github.com/odoo/odoo/blob/f033f94a029bc0455380259f1a7fc8489dc863e7/addons/pos_online_payment/controllers/payment_portal.py#L39)
return request.redirect(f'/web/login?redirect={urllib.parse.quote(request.httprequest.full_path)}')
Line may contain security risk: urllib.
    * [odoo/addons/pos_online_payment/controllers/payment_portal.py:72](https://github.com/odoo/odoo/blob/f033f94a029bc0455380259f1a7fc8489dc863e7/addons/pos_online_payment/controllers/payment_portal.py#L72)
return f'&exit_route={urllib.parse.quote(exit_route)}' if exit_route else ''
Line may contain security risk: urllib.

This method urllib.parse is used to quote the current URL to redirect to the login page or to quote the exit_route URL, which is used to allow the user to leave the payment portal page (for example in self-order, the payment portal page could be exited clicking on a button that opens this exit_route).

* [odoo/addons/pos_online_payment/models/pos_order.py:68](https://github.com/odoo/odoo/blob/f033f94a029bc0455380259f1a7fc8489dc863e7/addons/pos_online_payment/models/pos_order.py#L68)
self.env['bus.bus']._sendone(self.session_id._get_bus_channel_name(), 'ONLINE_PAYMENTS_NOTIFICATION', {'id': self.id})
Line may contain security risk: _sendone(.

The bus service is used to notify the POS frontend (cashier UI) when a new online payment is successful. No sensitive data is sent through the bus service. The bus channel is protected with the id and the secret uuid4 of the pos.session in his name, sent to the POS user when the session is loaded. This bus service notification ONLINE_PAYMENTS_NOTIFICATION triggers the web client to do a safe RPC to the server to get sensitive data (online payment state of the order).

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+    def pos_order_pay_transaction(self, pos_order_id=None, access_token=None, **kwargs):
+        """ Behaves like payment.PaymentPortal.payment_transaction but for POS online payment.
+
+        :param str pos_order_id: The POS order to pay, as a `pos.order` id
```
Sorry for the bad copy/paste in the doc. For the default value I didn't know if it was correctly handled by the web API.

thvi-odoo commented on addons/point_of_sale/models/pos_session.py:
```diff
@@ -1757,13 +1758,24 @@ def _get_pos_ui_account_tax(self, params):
+    def _ensure_access_token(self):
+        # Code taken from addons/portal/models/portal_mixin.py
+        if not self.access_token:
+            # we use a `write` to force the cache clearing otherwise `return self.access_token` will return False
```
I don't know, I'm only copy/pasting it from the current code of portal_mixin.

thvi-odoo commented on addons/point_of_sale/models/pos_order.py:
```diff
@@ -165,6 +169,10 @@ def _process_order(self, order, draft, existing_order):
         return pos_order.id
 
+    def _clean_payment_lines(self):
+        self.ensure_one()
```
This new method is an extraction of code that was done in the process of orders (one by one), needed to be overriden by the new module to manage online payments.
Of course the method could be changed to handle several orders in self, but it is currently not useful. I prefer to don't add complexity to a codebase that is already complex (by complex I don't mean the change would be difficult, we can just replace the '=' by 'in' and self.id by self.ids, I mean complex in understanding).
I personally don't like the "implied by the attribute access" principle, because it means that anyone that wants to use that method needs to read the full implementation to check if the self can hold only one record or several. Writing "self.ensure_one()" at the beginning of the method avoids that research.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+
+    @staticmethod
+    def _is_valid_currency(currency):
+        return currency and currency.active
```
Yes, it is **currently** just that. If tomorrow we need to add a check, it will be really easy: just one method to change.
Added to that, the method highlights and guarantees that the currency will be checked the same way in pos_order_pay and pos_order_pay_transaction.
So I disagree with you, this function is not useless.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+    def _get_amount_to_pay(order_to_pay_sudo):
+        if order_to_pay_sudo.state in ('paid', 'done', 'invoiced'):
+            return 0.0
+        amount = order_to_pay_sudo._get_checked_next_online_payment_amount()
+        if amount and PaymentPortal._is_valid_amount(amount, order_to_pay_sudo.currency_id):
+            return amount
+        else:
+            return order_to_pay_sudo.get_amount_unpaid()
```
You are considering the implementation of the method _is_valid_amount forgiving his role: to guarantee that the amount is valid for the payment portal page.
Whatever the _get_checked_next_online_payment_amount is good or not for the order point of view, it shouldn't, in my opinion, bypass the check of the payment portal.
A payment page is something dangerous in my opinion, I prefer to add more checks than needed, than reducing checks and create bypasses as you suggest.

thvi-odoo replied:
That's your point of view, I have a different one. I prefer to split responsibilities, you prefer to mix everything, allow everyone to change everything and add some checks in some cases only. I consider this dangerous and I try to avoid it.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+            return self._render_pay(request, rendering_context)
+        rendering_context['currency'] = currency_id
+
+        amount_to_pay = self._get_amount_to_pay(pos_order_sudo)
+        if not self._is_valid_amount(amount_to_pay, currency_id):
```
I'm not sure to understand, but you would like that _get_amout_to_pay returns the amount if it is valid, or something like False otherwise ?
So I would need to check if the result is a boolean, and check if it is False to determine if the amount is invalid and then display it on the payment page.
I find my approach easier to understand, it clearly highlights that I get the amount that should be paid, and then I check if it is valid for the payment portal page.

thvi-odoo replied:
I totally agree with you, except one thing: this is not a method intended to be used outside of the payment portal. This is an internal method only intended for the payment portal. It is a method to highlight that is is used several times, and guarantee that it will always produce the same result.
If you want the amount to pay of an order, you can call the method in pos.order. Nobody should and would use the payment portal to get the amount to pay of an order.

thvi-odoo commented on addons/pos_online_payment/models/payment_transaction.py:
```diff
@@ -0,0 +1,91 @@
+
+                if not tx.payment_id: # the payment could already have been created by account_payment module
+                    tx._create_payment()
+                if not tx.payment_id:
+                    raise ValidationError(
+                        "The POS online payment (tx.id=%d) could not be saved correctly" % tx.id)
```
So it is impossible that the creation of the payment fails ?
If it's the case, then this error is not a problem. But maybe one day there will be a situation where this won't be guaranteed, and I would be really happy to have that security that prevents to save a payment on the order that is not correctly saved.

thvi-odoo commented on addons/pos_online_payment/models/pos_payment_method.py:
```diff
@@ -0,0 +1,122 @@
+    has_an_online_payment_provider = fields.Boolean(compute='_compute_has_an_online_payment_provider', readonly=True)
+    type = fields.Selection(selection_add=[('online', 'Online')])
+
+    @api.depends('is_online_payment', 'journal_id', 'split_transactions')
```
Ok, good to know, thanks

thvi-odoo commented on addons/pos_online_payment/models/pos_payment_method.py:
```diff
@@ -0,0 +1,122 @@
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('is_online_payment', False):
```
Yes, but writting it explicitly shows what I consider: the field "is_online_payment" has a default value of False. Maybe tomorrow that will be changed, and the person who changes it will directly see it.
So yes it's not necessary, but I avoid a conversion of None to False, and I highlight what I consider.

thvi-odoo replied:
> Which matters how?
If the person who changes the default value of is_online_payment and forget to change it in this part of the code, anyone after can spot the issue, understand that it was False before and fix the issue. Otherwise there is the doubt of knowing if the person who wrote "vals.get('is_online_payment')" thought about the default value of the field.

> You're not actually avoiding anything useful. Conditions on None are idiomatic, it's special-cased in all sorts of locations in the interpreter. You're saving a C-level branch, at the cost of adding an instruction (a much larger branching), an array lookup, and increased refcount traffic.
I perfer to don't waste my time on considering performances of such little things. The conversion is mental, the code is meant to be read.

thvi-odoo commented on addons/pos_online_payment/models/pos_payment_method.py:
```diff
@@ -0,0 +1,122 @@
+    def _force_online_payment_values(vals, if_present=False):
+        if 'type' in vals:
+            vals['type'] = 'online'
+        if not if_present or 'split_transactions' in vals:
+            vals['split_transactions'] = False
+        if not if_present or 'receivable_account_id' in vals:
+            vals['receivable_account_id'] = False
+        if not if_present or 'outstanding_account_id' in vals:
+            vals['outstanding_account_id'] = False
+        if not if_present or 'journal_id' in vals:
+            vals['journal_id'] = False
+        if not if_present or 'is_cash_count' in vals:
+            vals['is_cash_count'] = False
+        if not if_present or 'use_payment_terminal' in vals:
+            vals['use_payment_terminal'] = False
```
Thank you for the idea

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+        if not partner_sudo:
+            return self._redirect_login(request)
+
+        kwargs = {
+            'pos_order_id': pos_order_sudo.id,
+        }
```
Maybe today, but tomorrow if you want to add something, you just need to add it to this dict.
I try to use as little copy/paste I can, because duplicating stuff usually ends up with incoherent code, particularly in a community project as Odoo where everyone doesn't take the same care when changing something.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+            **self._get_custom_rendering_context_values(**kwargs),
+        }
+
+        currency_id = pos_order_sudo.currency_id
+
+        if not self._is_valid_currency(currency_id):
```
I disagree with that approach, as explained earlier.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+                    _("Tokenization is not available for logged off customers."))
+            kwargs['custom_create_values']['tokenize'] = False
+
+        currency_id = pos_order_sudo.currency_id
+        if not self._is_valid_currency(currency_id):
```
Same answer

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,299 @@
+        if not payment_option_id:
+            raise UserError(_("A payment option must be specified."))
+        flow = kwargs.get('flow', None)
+        if not (flow and flow in ['redirect', 'direct', 'token']):
```
There is a battle on that. Joseph wanted:
`if not (flow and flow in ['redirect', 'direct', 'token']):`
I prefer:
`if not flow or flow not in ['redirect', 'direct', 'token']:`
And what you propose potentially allows the "None" value to be put in the array. It's a detail but you are checking if None is not in the array, looping on it.

thvi-odoo replied:
We certainly don't want "None" to be considered as a valid value. We are in a payment processing, we need to have a valid and clear flow, and clearly don't want to guess it.
I mentioned it was a detail to highlight the fact that it's not a performance issue, more a readable issue. I prefer to distinguish the "None" case as a special case. You prefer to mix it as a normal value, which I consider dangerous but it's only my point of view.

thvi-odoo commented on addons/pos_online_payment/models/pos_payment.py:
```diff
@@ -0,0 +1,41 @@
+    def create(self, vals_list):
+        for vals in vals_list:
+            if 'payment_method_id' in vals:
+                payment_method = self.env['pos.payment.method'].browse(vals['payment_method_id'])
+                if payment_method.exists() and payment_method.is_online_payment:
+                    online_account_payment_id = vals.get('online_account_payment_id')
+                    if not online_account_payment_id or not self.env['account.payment'].browse(online_account_payment_id).exists():
+                        raise UserError("Cannot create a POS online payment without an accounting payment.")
```
Ok, done

thvi-odoo commented on addons/pos_online_payment/models/pos_payment.py:
```diff
@@ -0,0 +1,41 @@
+        return super().create(vals_list)
+
+    def write(self, vals):
+        if any((payment.online_account_payment_id or payment.payment_method_id.is_online_payment) and any(k in ('amount', 'payment_date', 'payment_method_id', 'online_account_payment_id', 'pos_order_id') for k in vals.keys()) for payment in self):
```
Ok, thanks

thvi-odoo commented:
> Why is everything a staticmethod? They're completely useless, just use free functions.

I removed some staticmethod decorators, but kept them for the functions that could be accessed from outside of the payment controller (to get an URL for example).
My intention was to not have "self" argument in functions where "self" is not used. But since you told me that we have access to "request" variable from those methods, I find it more clear to have those methods not static since I removed their "request" parameter.

caburj commented:
Hi @xmo-odoo, @thvi-odoo tried resolving all your comments 2 days ago. Then we both go thru each of it this morning and we believe we have answered your questions. Kindly check this PR again and see if we can get the security override.

caburj commented:
Hi @xmo-odoo. Kindly check again if we can get a security override.

Hi @thvi-odoo. Kindly label the commits that concern XMO's comments so that he can easily find how you resolved his comments.

thvi-odoo commented:
I made the changes for XMO's comments in the "temp4" (https://github.com/odoo/odoo/pull/123237/commits/690b98c8858f49341763a20d77ab13856863776c) commit.

caburj reviewed (approved):
robodoo r+

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,291 @@
+
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login()
```
You are right, this code should never be executed. But because payments are a sensitive thing, I prefer to add this small check and prevent any issue. It's again the same principle: I try to check everything I can because I don't want to trust things coming from outside.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,291 @@
+        """
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+        self._ensure_session_open(pos_order_sudo)
+        exit_route = request.httprequest.args.get('exit_route')
```
I think (I wrote this a long time ago) kwargs are data coming from the JS code of the 'payment' module, and the exit_route is not in that kind of data, it is only in the URL args (and it is optional). I didn't wanted to change the JS code of the 'payment' module, and found this easy way of doing what I needed.

thvi-odoo commented on addons/pos_online_payment/controllers/payment_portal.py:
```diff
@@ -0,0 +1,291 @@
+        exit_route = request.httprequest.args.get('exit_route')
+        user_sudo = request.env.user
+        logged_in = not user_sudo._is_public()
+        partner_sudo = self._get_partner_sudo(user_sudo)
+        if not partner_sudo:
+            return self._redirect_login()
```
Same answer.

AntoineVDV commented on addons/pos_online_payment/models/payment_transaction.py:
```diff
@@ -0,0 +1,85 @@
+
+    def _set_authorized(self, state_message=None, **kwargs):
+        """ Override of payment to process POS online payments automatically. """
+        super()._set_authorized(state_message=state_message, **kwargs)
```
It is a very bad idea to perform complex processing at this point because a server error will trigger a cursor rollback which will undo the creation of the transaction. You'll lose track of the payment, but it will still have been made on the external system.

Let `_reconcile_after_done` do the heavy-lifting instead.

thvi-odoo replied:
Ok thank you for the information. So it is impossible to have a transaction where _set_authorized would be called and _reconcile_after_done would never be called (just to be sure) ?

thvi-odoo replied:
The goal is to detect when a transaction is a successful payment to make the POS order as paid. Good to know _reconcile_after_done is enough to do that.

thvi-odoo replied:
Hello @AntoineVDV, I looked at the code and the sale module also overrides _set_authorized, why would POS would process payments differently than sale ?

AntoineVDV replied:
This kind of override is being removed from all apps as we speak. See https://github.com/odoo/enterprise/commit/126f0087dd8c9dbd9de396a5a14ae05f2a091fba for example.

The override in `sale` will be removed too, but it's far from being as problematic as the one in `pos_online_payment`, which creates `account.payment` records. The latter is the easiest way to make sure some payments will disappear into thin air every now and then.

thvi-odoo replied:
Ok thank you

Xavier-Do commented on addons/pos_online_payment/tests/test_frontend.py:
```diff
@@ -0,0 +1,216 @@
+            expected_payment_provider_id = self.payment_provider.id
+            if not isinstance(next_online_payment_amount, bool) and not tools.float_is_zero(next_online_payment_amount, precision_rounding=method_self.currency_id.rounding) and next_online_payment_amount > 0:
+                # Delay must be long enough to execute _fake_online_payment after the current RPC call is answered
+                t = threading.Timer(5, self._fake_online_payment, args=(pos_order_id, order_access_token, expected_payment_provider_id))
```
This is a bad idea to spawn a thread inside a test, it is a very good way to introduce randomness. 

The request that are triggered inside _fake_online_payment should have a test cursor and be sequential with the other request, but any call to self.env.cr will be concurrent and can lead to concurrent access to the cursor. 
 
 As an example _get_tx will make a search, so will execute and fetch on the same cr as a request, and concurrently.
 
 This can lead to error like in https://runbot.odoo.com/runbot/build/54372105
 
 ```
   File "/data/build/odoo/odoo/models.py", line 3597, in _fetch_query
    rows = self.env.cr.fetchall()
psycopg2.ProgrammingError: no results to fetch
```
This should be done sequentially. It is hard to do since inside a request, the thread will deadlock trying to generate another request. This is why the timout is needed here.

A solution would be to split the tour in parts, start the tour, at the end, simulate the answer, then start the rest of the tour. 
Another solution is to simulate the answer without making a request, by calling the code inside the endpoint directly 
A last proposition is to trigger the endpoint response from the tour if it makes sense.

thvi-odoo replied:
Hello @Xavier-Do, thank you for your explanations and your suggestions.

The idea of this test is to simulate an almost real flow: the cashier initiates an online payment, the server receives it and answers normally to the cashier, but the server also uses the received request to generate a fake online payment (by doing web requests to the online payment page, as a normal customer would do to pay). The cashier then receives this fake online payment and checks it.
(Because websocket communication doesn't seem to work in the test environment, I had to make the JS tour web client frequently request the state of the (fake) online payment from the server)

I'm not sure to understand what you wrote, could you confirm me the following please ?

> The request that are triggered inside _fake_online_payment

By "request", I guess you are talking about _fake_open_pos_order_pay_page, _fake_request_pos_order_pay_transaction_page and _fake_open_pos_order_pay_confirmation_page.

> should have a test cursor and be sequential with the other request

Are you describing the current code ("should" used for an assumption) or are you asking for that to be done ("should" used to express a condition to follow to make the test correctly work) ?
By "other request", are you talking about _fake_get_and_set_online_payments_data ?

> but any call to self.env.cr will be concurrent and can lead to concurrent access to the cursor.

Any call from where ? During the execution of the thread for _fake_online_payment ?

> As an example _get_tx will make a search, so will execute and fetch on the same cr as a request, and concurrently.

Is the "request" in your sentence referring to any request or a special request in the test ?
What I understand is that since _fake_online_payment is executed in a separate thread, and because it executes _get_tx which does a search with the cursor, another thread could also do a request to the same cursor at the same time, concurrently.

> This should be done sequentially.

By "This" I assume you are talking about _fake_online_payment execution.

> It is hard to do since inside a request

By "request" I assume you are talking about the RPC to _fake_get_and_set_online_payments_data.

> the thread will deadlock trying to generate another request. This is why the timout is needed here.

I don't understand what timeout you are talking about.
I understand that the thread executing the RPC to _fake_get_and_set_online_payments_data would deadlock with the web requests in _fake_online_payment if _fake_online_payment was called synchronously in _fake_get_and_set_online_payments_data.
I think that it was my first approach, but indeed it seems that doing web communications inside an RPC is not supported.

I would like to know, is the problem caused by the test itself (depending on how the 2 threads (the one for the RPC to _fake_get_and_set_online_payments_data and the one for executing _fake_online_payment) of the test are executed (slow/fast execution because of busy test server for example ?), they can concurrently use the same cursor ?), or the way it is executed (are several tests executed simultaneously for example, and inducing the random error ?).

> A solution would be to split the tour in parts, start the tour, at the end, simulate the answer, then start the rest of the tour.

What I understand is that you suggest to have 2 tours, and between those 2 tours simulate the answer.
From my understanding, this is not possible and wouldn't represent a realistic flow because the web client would lose the online payment popup (showing the current online payment, and waiting for the successful online payment notification from the server). It wouldn't be possible to open a second tour with the state of the online payment popup displayed and all the JS code ready to receive the answer correctly (of course everything is possible, but from my point of view it would need a lot of hacking code specific to the test, that at the end wouldn't test the actual working of pos_online_payment).

> Another solution is to simulate the answer without making a request, by calling the code inside the endpoint directly

I understand that you suggest to directly call pos_order_pay, pos_order_pay_transaction and pos_order_pay_confirmation of payment_portal.py from the _fake_online_payment implementation (and call _fake_online_payment synchronously in _fake_get_and_set_online_payments_data).
How would the "request" variable be filled ? Wouldn't it cause issues to execute controller methods like that (from the execution of the RPC to _fake_get_and_set_online_payments_data) ?

> A last proposition is to trigger the endpoint response from the tour if it makes sense.

I'm sorry I don't understand. That's already what is done: the tour is triggering the server to do the fake requests to the online payment page (with the pos_order_id and order_access_token that are generated during the tour and are received from it).
The idea of the test is to test a full flow: the cashier interface and the customer payment web interface. If the web requests to the payment page are removed, the test would become significantly less interesting.

I would have a last question please: Is it possible to schedule the execution of some code (_fake_online_payment) after some delay on the main thread (and not a separate one as I did) during the execution of a tour ?
If it's not possible, then I think that I will just simplify the test.
Thank you in advance for your explanations.
